### PR TITLE
T5 Encoder

### PR DESCRIPTION
**Pull request by calvinpelletier**
#### Context
What is the purpose of this PR? Is it to
- [x] add a new feature
- [ ] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

Please link to any issues this PR addresses.

#### Changelog
What are the changes made in this PR?
* T5 tokenizer
* T5 encoder
* convert weights from HF's T5 to ours
* unit tests

#### Analysis

Comparison to HF's implemention (batch of text -> encoder output):
- 6.7e-5 MSE output difference
- ours is ~5% faster

#### Test plan
Please make sure to do each of the following if applicable to your PR. If you're unsure about any one of these just ask and we will happily help. We also have a [contributing page](https://github.com/pytorch/torchtune/blob/main/CONTRIBUTING.md) for some guidance on contributing.

- [x] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [x] add [unit tests](https://github.com/pytorch/torchtune/tree/main/tests/torchtune) for any new functionality
- [x] update [docstrings](https://github.com/pytorch/torchtune/tree/main/docs/source) for any new or updated methods or classes
- [x] run unit tests via `pytest tests`
- [x] run recipe tests via `pytest tests -m integration_test`
- [x] manually run any new or modified recipes with sufficient proof of correctness
- [x] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

#### UX
If your function changed a public API, please add a dummy example of what the user experience will look like when calling it.
Here is a [docstring example](https://github.com/pytorch/torchtune/blob/6a7951f1cdd0b56a9746ef5935106989415f50e3/torchtune/modules/vision_transformer.py#L285)
and a [tutorial example](https://pytorch.org/torchtune/main/tutorials/qat_finetune.html#applying-qat-to-llama3-models)

- [x] I did not change any public API
- [ ] I have added an example to docs or docstrings

#### Minimal test code

```python
import torch

from torchtune.models.t5 import t5_tokenizer, t5_v1p1_xxl_encoder
from torchtune.training.checkpointing._checkpointer import FullModelHFCheckpointer

MAX_SEQ_LEN = 512

# tune download google/t5-v1_1-xxl --output-dir /tmp/t5-hf
tokenizer = t5_tokenizer("/tmp/t5-hf/spiece.model", max_seq_len=MAX_SEQ_LEN)
checkpointer = FullModelHFCheckpointer(
    "/tmp/t5-hf",
    ["pytorch_model.bin"],
    "T5_ENCODER",
    "/tmp/t5-tt",
)

model = t5_v1p1_xxl_encoder(max_seq_len=MAX_SEQ_LEN)
model.load_state_dict(checkpointer.load_checkpoint()["model"])
model = model.to(device="cuda", dtype=torch.bfloat16).eval().requires_grad_(False)


def tokenize(texts):
    result = torch.full(
        (len(texts), tokenizer.max_seq_len),
        tokenizer.pad_id,
        dtype=torch.int,
    )
    for i, text in enumerate(texts):
        tokens = tokenizer.encode(text)
        result[i, : len(tokens)] = torch.tensor(tokens)
    return result


tokens = tokenize(
    [
        "a cow jumping over the moon",
        "a helpful AI assistant",
    ]
)
encoding = model(tokens.to("cuda"))
```

### CODE REVIEW

File: torchtune/models/t5/_encoder.py [hunk 1]
```diff
@@ -0,0 +1,304 @@
...
+        # [bsz, seq_len, embed_dim] -> [bsz, num_heads, seq_len, head_dim]
+        q = q.view(bsz, seq_len, self.num_heads, self.head_dim).transpose(1, 2)
+        k = k.view(bsz, seq_len, self.num_heads, self.head_dim).transpose(1, 2)
+        v = v.view(bsz, seq_len, self.num_heads, self.head_dim).transpose(1, 2)
+
+        # attention with relative position bias
+        attn_score = torch.matmul(q, k.transpose(-2, -1))
+        attn_score += rel_pos_bias
+        attn_weight = F.softmax(attn_score.float(), dim=-1).to(attn_score.dtype)
+        attn_out = torch.matmul(attn_weight, v)
```
Comment (calvinpelletier) [inline torchtune/models/t5/_encoder.py:209]: This part could be simplified by using `F.scaled_dot_product_attention` by repurposing the `mask` argument for `rel_pos_bias` (because `scaled_dot_product_attention` simply adds the mask to the attention score when the mask is a float tensor). However, when I tried this it was significantly slower for some reason

File: torchtune/models/t5/_encoder.py [hunk 2]
```diff
@@ -0,0 +1,304 @@
...
+            torch.Tensor: relative position bias tensor with shape [1, num_heads, max_seq_len, max_seq_len]
+        """
+        # convert bucket numbers to bias values for each attention head
+        x = self.embedding(self.relative_position_to_bucket)
+
+        # shape [max_seq_len, max_seq_len, num_heads] -> [1, num_heads, max_seq_len, max_seq_len]
+        return x.permute([2, 0, 1]).unsqueeze(0)
+
+
+def _calc_birectional_rel_pos_to_bucket(
```
Comment (calvinpelletier) [inline torchtune/models/t5/_encoder.py:258]: should I add more comments in this function explaining each operation? or is it fine to just leave it a bit opaque

File: torchtune/models/t5/_model_builders.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+from torchtune.models.t5._component_builders import t5_encoder
+from torchtune.models.t5._encoder import T5Encoder
+from torchtune.models.t5._tokenizer import T5Tokenizer
+
+
+def t5_v1p1_xxl_encoder(max_seq_len: int = 512) -> T5Encoder:
```
Comment (calvinpelletier) [inline torchtune/models/t5/_model_builders.py:12]: thoughts on writing decimal points as `p` instead of `_` in snake case? imo it's hard to read the `_` decimals:

`t5_v1_1_xxl_encoder` vs `t5_v1p1_xxl_encoder`

File: torchtune/models/t5/_encoder.py (hunk 1 above)
Comment (ebsmothers) [reply torchtune/models/t5/_encoder.py:209]: I wonder whether this is a case where we could benefit from using flex attention?

**Review by pbontrager — COMMENTED** [review]
I gave this a high level pass so far. This looks really clean and good. My only concern is wish us having to have a custom T5 layer and attention module. If flex attention would let us use our existing modules I'd prefer to go down that route.

File: torchtune/modules/tokenizers/_sentencepiece.py [hunk 4]
```diff
@@ -7,6 +7,7 @@
 from typing import List, Optional
 
 from sentencepiece import SentencePieceProcessor
+
```
[the remark this answers is not in the document]
Comment (calvinpelletier) [reply torchtune/modules/tokenizers/_sentencepiece.py:10]: its from the linter

File: torchtune/models/t5/_model_builders.py (hunk 3 above)
Comment (pbontrager) [reply torchtune/models/t5/_model_builders.py:12]: I think the p could be better. But I'd want to switch everything to this notation then

File: torchtune/models/t5/_encoder.py (hunk 1 above)
Comment (calvinpelletier) [reply torchtune/models/t5/_encoder.py:209]: Flex attention requires boolean masks AFAICT: https://github.com/pytorch/pytorch/blob/main/torch/nn/attention/flex_attention.py#L826-L827

File: torchtune/models/t5/_tokenizer.py [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+        max_seq_len (int): the context length
+        truncate (bool): whether to truncate the token sequence when longer than max_seq_len
+    """
+
+    def __init__(self, path: str, max_seq_len: int = 512, truncate: bool = True):
+        super().__init__(path)
+        self.max_seq_len = max_seq_len
+        self.truncate = truncate
+
+    def encode(self, text: str) -> List[int]:
```
[the remark this answers is not in the document]
Comment (calvinpelletier) [reply torchtune/models/t5/_tokenizer.py:26]: It has it. It's in the base tokenizer class (`SentencePieceBaseTokenizer`). It just didn't require any modifications

File: torchtune/models/t5/_encoder.py [hunk 6]
```diff
@@ -0,0 +1,304 @@
...
+        sa_norm: nn.Module,
+        mlp_norm: nn.Module,
+    ) -> None:
+        super().__init__()
+        self.attn = attn
+        self.mlp = mlp
+        self.sa_norm = sa_norm
+        self.mlp_norm = mlp_norm
+
+    def forward(self, x: Tensor, rel_pos_bias: Tensor) -> Tensor:
```
[the remark this answers is not in the document]
Comment (calvinpelletier) [reply torchtune/models/t5/_encoder.py:116]: Well, kind of... it's a float tensor that gets added to the attention scores (which is how `scaled_dot_product_attention` deals with float masks, so yeah we could think of it as a mask). 

I didn't use our modules because: 
- we can't use flex attention with float masks
- our `MultiHeadAttention`/`TransformerSelfAttentionLayer` modules expect the masks to be boolean tensors (according to the docstring at least)
- `MultiHeadAttention` uses the default attention scaling (`1/sqrt(dim)`), but T5 doesn't scale it at all
- `MultiHeadAttention` uses `F.scaled_dot_product_attention` which is much slower with float masks than the manual implementation I went with

We could switch to our modules with a couple small changes (update the docstrings to clarify that the mask can also be a float tensor and add an argument for disabling attention scaling), but given how little code is required to just implement separate versions for T5, I thought it was cleaner to leave our modules alone (especially since this implementation is faster).

File: torchtune/models/t5/_model_builders.py (hunk 3 above)
Comment (calvinpelletier) [reply torchtune/models/t5/_model_builders.py:12]: I'll make a separate PR for this

Comment (pbontrager) [reply torchtune/models/t5/_model_builders.py:12]: Can you use the _ notation for now and just update this along with everything else in that PR?

File: torchtune/models/t5/_encoder.py (hunk 1 above)
Comment (pbontrager) [reply torchtune/models/t5/_encoder.py:209]: I believe you can use any mask for flexattention with scoremod. I think we should land this as is but create an issue to update this to use flex attention.

**Review by pbontrager — APPROVED** [review]
I think this looks good. I'd like an issue to update the attention to use flex attention but that doesn't need to happen here.
